### ISSUE TITLE
xSQLServerAlwaysOnAvailabilityGroupDatabaseMembership: Logic error in tests

### ISSUE THREAD

**Issue by johlju** [PowerShell/xSQLServer#673]
**Details of the scenario you tried and the problem that is occurring:**
When trying to opt-in for common tests, the tests for xSQLServerAlwaysOnAvailabilityGroupDatabaseMembership started to fail..

See errors:
https://ci.appveyor.com/project/johlju/xsqlserver/build/6.0.700.0#L1717

```plaintext
System.InvalidOperationException: Neither PowerShell module SqlServer or SQLPS was found. Unable to run SQL Server cmdlets.
        at <ScriptBlock>, C:\projects\xsqlserver\DscResources\CommonResourceHelper.psm1: line 88
        88:     throw $errorRecordToThrow
```

I think it has to do with this row that loads Import-SQLPSModule in the construct on initialization.
```powershell
    # Load the SQLPS module when the class is initialized
    xSQLServerAlwaysOnAvailabilityGroupDatabaseMembership()
    {
        Import-SQLPSModule
    }
```

That means that this code in the tests fail, because the mocks are only active when the It-block is run, not when BeforeEach is run. 

```powershell
            BeforeEach {
                $databaseMembershipClass = [xSQLServerAlwaysOnAvailabilityGroupDatabaseMembership]::New()
                $databaseMembershipClass.DatabaseName = $mockDatabaseNameParameter.Clone()
                $databaseMembershipClass.SqlServer = 'Server1'
                $databaseMembershipClass.SQLInstanceName = 'MSSQLSERVER'
                $databaseMembershipClass.AvailabilityGroupName = 'AvailabilityGroup1'
            }
```

The downside of having that in the construct is that each call to initialize a new object will call Import-SQLPSModule regardless if it is needed or not.
```powershell
$currentConfiguration = New-Object xSQLServerAlwaysOnAvailabilityGroupDatabaseMembership 
```

But the above seems wrong to, because then the tests would work at all. But they do, unless I opt-in for the common tests. What that does is activate markdown checking and example compiling. And I think it is the last that changes this behavior in some way. 

BUT, I can repeat this error in the command line. When running the test in a new PowerShell window, the tests works, but if I run it a second time, the test fail with the same error as above. If I change the code, like just a white space, then the test runs fine on a third try, but when I then run a forth time it fails again, and so on. Which make me conclude that something fishy is going on with Pester and class-based resources.

**The DSC configuration that is using the resource (as detailed as possible):**
n/a

**Version of the Operating System, SQL Server and PowerShell the DSC Target Node is running:**
n/a

**What module (SqlServer or SQLPS) and which version of the module the DSC Target Node is running:**
n/a

**Version of the DSC module you're using, or 'dev' if you're using current dev branch:**
Dev

**Comment by johlju** [comment]
@randomnote1 Do you have any thoughts on this one? Did you experience this too when coding this one? 

Changing this row in appveyor.yml triggers this error.

```powershell
Invoke-AppveyorTestScriptTask -CodeCoverage -CodeCovIo
```

to this row 

```powershell
Invoke-AppveyorTestScriptTask -CodeCoverage -CodeCovIo -ExcludeTag @()
```

**Comment by randomnote1** [comment]
I'm not sure about the AppVeyor stuff since AppVeyor is still somewhat of a black box to me. I do see errors when I attempt to execute the pester tests twice in a row in the same session.

`Invoke-Pester -Script .\Tests\Unit\MSFT_xSQLServerAlwaysOnAvailabilityGroupDatabaseMembership.Tests.ps1`

I normally end up calling a fresh session each time in order to ensure the SMO stubs load cleanly.

`powershell.exe -Command { Invoke-Pester -Script .\Tests\Unit\MSFT_xSQLServerAlwaysOnAvailabilityGroupDatabaseMembership.Tests.ps1 }`

**Comment by randomnote1** [comment]
I think I put the `Import-SQLPSModule` in the class load because at one point I had passed typed objects into the methods of the class. Since that is no longer the case, it can probably be moved to locations where it is required.

**Comment by johlju** [comment]
@randomnote1 the problem doesn't seem to be SMO stubs, but rather that the mock's doesn't trigger for Import-SQLPSModule (or it is related to SMO stubs). Since the mock isn't called the actual function is called, and it doesn't find any of the modules because they doesn't exist on the AppVeyor build worker (nor my client).

I see this row in the module manifest. Is that necessary for the loading of the module? 

```
RootModule = '.\DSCResources\MSFT_xSQLServerAlwaysOnAvailabilityGroupDatabaseMembership\MSFT_xSQLServerAlwaysOnAvailabilityGroupDatabaseMembership.psm1'
```

The explanation below says that setting this makes this a Script module, and not a manifest module.

>Possible types for the root module can be empty (which will make this a Manifest module), the name of a script module (.psm1, which makes this a Script module), or the name of a binary module (.exe or .dll, which makes this a Binary module). Placing the name of a module manifest (.psd1) or a script file (.ps1) in this element will cause an error to occur.
https://msdn.microsoft.com/en-us/library/dd878337(v=vs.85).aspx

**Comment by johlju** [comment]
@randomnote1  The resource can not be found when the row RootModule is removed. How will this work when the next class-based resource is added? 

I feel now that we should pull this resource before the release and look over this once again. It doesn't seem that this is correctly implemented.
I don't want to release this resource potentially making a breaking change down the road.

**Comment by johlju** [comment]
@randomnote1 I will send in a PR pulling this resource. I will send in another PR with the changes that was made with this resource so we can continue work on this an figure out how this should be implemented.

**Comment by johlju** [comment]
@kwirkykat or @mbreakey3 I need some guidance. I feel that I need to pull the class-based resource xSQLServerAlwaysOnAvailabilityGroupDatabaseMembership from xSQLServer before release. I don't feel this has been implemented correctly. Or there are bugs around how class-based resource as loaded together with mof-based resource. If you see any thing we can do to keep this in then let me know. 
 (the need to add it to RootModule property in the module manifest for it to be detected by Get-DscResource).

We could keep this in by converting it to a mof-based resource. But don't think we have time to do that before release.

@randomnote1 You are the author of this resource, what do you say?

**Comment by johlju** [comment]
@randomnote1 I deleted previous comment about sending in PR's to pulling this resource. I will not do that yet. But I do have prepared a PR to pull this resource; https://github.com/johlju/xSQLServer/tree/pull-resource-xSQLServerAlwaysOnAvailabilityGroupDatabaseMembership
And I have prepared a PR to add back this resource; https://github.com/johlju/xSQLServer/tree/pull-resource-xSQLServerAlwaysOnAvailabilityGroupDatabaseMembership

I hope we can find another solution though. We could keep it in, and change it to a mof based resource for the next release, I'm just afraid that the current implementation could break for user since it seems to have an odd behavior.

**Comment by randomnote1** [comment]
Let me know what you decide. The problem with the class based resource is that it requires that line in the psd1. It took me a while to figure that out 🙁. I'll be back in the office tomorrow, so I'll comment more then.

**Comment by johlju** [comment]
It seems very strange that it is needed, or it should point to something else that can load more than one class based resource. 

Beside Katie (also pinged her on Twitter) and Mariah, I have asked @PlagueHO and @bgelens (pinged thru Twitter) to see if they can help with any suggestions here. :)

Any if anyone in the community has any input on this, feel free to come with suggestions 😃

**Comment by PlagueHO** [comment]
Hi @johlju - I've not actually implemented a class based resource before (I tend to avoid them due to all the issues we've had in the past with testing and compatibility with PS 4.0).

Question (excuse my ignorance): Why is the `Import-SQLPSModule` being called in the class constructor instead of the Set/Test/Get functions like it is in the MOF based resources?

**Comment by johlju** [comment]
I had a conversation with @bgelens on Twitter regarding this issue this morning. He come with a few suggestions that might help, especially with the RootModule problem. I really appreciate Ben taking his time from his busy schedule going thru this! Thank you!

This is a summary of what he said.

In his experience 
*  The class-based resource needs to have it's own manifest and as it's rootmodule just the base name of the psm1 (no other keys required). Then, the own resource manifest needs to be in the NestedModules property of the parent module.
  _Regarding the base name, my guess is that Ben means no MSFT\_ prefix on anything. That is also said in the specific contribution guideline (which I wrote with the help of Katie, and forgot until now); "Any composite (with a Configuration) or class-based resources should be prefixed with just 'xSQLServer'."._
* There is currently a bug with DscResources subfolder and class resources. When the class resource resides in DscResources, the auto-generated schema file will be rendered using the wrong module name. If placed in DscClassResources subfolder instead it works correctly.
  _It sounds like that this could be the problem we are experience, when Get-DscResource can't find the resource._

**Comment by johlju** [comment]
@PlagueHO it shouldn't be there, it should be in each of the Set/Test/Get functions. This is an old technical debt I missed in the review. @randomnote1 had that in at the beginning when he used specific SQL types from the imported module. These types was removed, but this was left in place since it works. But I think this could conflict with the test, and also it loads the SQL module even when you just initiate the class (empty class). @bgelens suggested in his quick glance at this that we could move it to the top of the resource module, but then we can mock it easily. 
So this is gonna be moved into each Set/Test/Get functions

**Comment by johlju** [comment]
@randomnote1 You are most familiar with this resource, so if you have time, it would be awesome f you could try out these ideas and send in a PR. 😄 Let me know if you have time, otherwise I will look at this.

**Comment by johlju** [comment]
@kwirkykat also made comments and got mof and class-based resource to work together in this comment https://github.com/PowerShell/xSQLServer/issues/425#issuecomment-298436623. So I guess we misunderstood how to make this work. @kwirkykat could you say, or even better, can you show what you did to get it to work? 😄

**Comment by TraGicCode** [comment]
One thing to keep in mind is if there is no MOF file then when this new resource officially gets released puppet users will not be able to utilize this dsc resource from the puppetlabs/puppetlabs-dsc puppet module.

**Comment by randomnote1** [comment]
@TraGicCode, Can you elaborate on Puppet? I have zero experience/knowledge about it.

**Comment by randomnote1** [comment]
@johlju, I'll try to catch up on comments here.

> The resource can not be found when the row RootModule is removed. How will this work when the next class-based resource is added?

This was one of my questions/concerns when submitting this that I forgot to bring up during the review process. I was hoping for a little more guidance from the PowerShell team on how to handle this. It was my understanding that all of the class-based resources needed to be built and/or referenced in a common file. Not really sure though.

> The class-based resource needs to have it's own manifest and as it's rootmodule just the base name of the psm1 (no other keys required). Then, the own resource manifest needs to be in the NestedModules property of the parent module.

I'm pretty sure I tried this in the early stages of testing, however I don't recall. I will test and report back.

> There is currently a bug with DscResources subfolder and class resources. When the class resource resides in DscResources, the auto-generated schema file will be rendered using the wrong module name. If placed in DscClassResources subfolder instead it works correctly.

This is interesting and sounds promising. Will test.

**Comment by jpogran** [comment]
@randomnote1 What @TraGicCode means is that currently the Puppet dsc module parses MOF files to  generate Puppet types. Since class based DSC Resources don't have MOF files, we (puppet) need to change our build tasks to read something else. We have ticketed work to address this soon.

**Comment by randomnote1** [comment]
@jpsnover, thanks for the explanation!

**Comment by jpogran** [comment]
Well, I can officially retire happy now. Being confused with @jpsnover is something I will never top. :)

**Comment by randomnote1** [comment]
lol...oops! A tad quick on the tab complete :)

**Comment by randomnote1** [comment]
@johlju, I was able to find some good information in the ServiceBusForWindowsServerDsc resource by referencing this issue: PowerShell/DscResources#208. It also looks like there's some good AppVeyor information in there for you to take a look at.

I was able to get the nested modules working as expected by placing a module manifest in the resource directory. Also, based on the ServiceBusForWindowsServerDsc, I think we should utilize the `DSCClassResources` directory for all class-based resources.

I'll submit a pull request based on this information and we can do our work in that.

**Comment by johlju** [comment]
@randomnote1 Awesome! Glad to hear that you made progress! 😄  I will take a look at the issue while you getting a PR together. We can hold of releasing this module a couple of days, if we need to, at least until we solved that we are able to add more class-based resources. Release is schedule 12 July if I remember correctly.

**Comment by randomnote1** [comment]
@johlju, I've identified a few issues that will prevent unit tests from working:

- In `DSCResource.Tests\TestHelper.psm1` starting at line 274, the `Initialize-TestEnvironment` function assumes all resources will live in `DSCResources`.
- The `DSCResources\CommonResourceHelper.psm1` file lives in the `DSCResources` folder and therefore breaks the import of localized data.
- Pester doesn't seem to want to load the module either 
> Module 'MSFT_xSQLServerAlwaysOnAvailabilityGroupDatabaseMembership' is not a Script module.

How would you like me to proceed? Submit a known broken PR in to test and collaborate? Or try and fix these issues siloed?

**Comment by johlju** [comment]
Problem 1 and 2 seems fixable, but how about the third problem? Is it because of the first problem?

What is you opinion on this. Is this worth putting time on fixing these problems rather than converting this resource to a mof resource? 

If you think it is worth puWe can collaborate on the same PR.

**Comment by johlju** [comment]
@randomnote1 Seems that we need to split the logic from the actual class resource to be able to test it properly with pester, after reading Katie's comment here https://github.com/PowerShell/DscResources/issues/208#issuecomment-255462702. So essentially creating Get-/Set-/Test-TargetResource functions in a psm1 file. if we put that .psm1 file under DscResource\xSQLServerAlwaysOnAvailabilityGroupDatabaseMembership and then we load that in the unit test, would that solve problem 1 and 3?

**Comment by randomnote1** [comment]
I think it's worth putting in the time because there have been one or two other folks who asked about submitting class-based resources in the past. I'll take a stab at creating the .psm1 file and see if that fixes our issues. Not sure how I missed that comment the first time around!

**Comment by johlju** [comment]
@randomnote1 Awesome! Let's get this done then. 😄 I would love to see that it is possible to run the pester test twice without the mocks breaking. 

Would it help if I send in a PR to DscResource.Tests to get problem 1 solved? Essentially the same code changes as here https://github.com/PowerShell/DscResource.Tests/pull/64.

**Comment by randomnote1** [comment]
@johlju, that would be great if you could do that! Moving the conversation over to Reviewable.

**Comment by johlju** [comment]
@randomnote1 Added a new parameter ResourceType to the test helper function Initialize-TestEnvironment to be able to test class-based resources in the folder DscClassResources. The new parameter ResourceType can be set to either 'Mof' or 'Class'. Default value for parameter ResourceType is 'Mof'.

I don't want to PR in it until it has been tested. 

Working branch: https://github.com/johlju/DscResource.Tests/tree/class-based

So when you got a DscClassResource folder, you can test is using the working branch by cloning it into xSQLServer folder (replacing the DscResource.Test folder).

To test it using AppVeyor change this line in the AppVeyor.yml.
```
- git clone https://github.com/PowerShell/DscResource.Tests
```
to
```
- git clone https://github.com/johlju/DscResource.Tests
- ps: Push-Location
- ps: cd "$env:APPVEYOR_BUILD_FOLDER\DscResource.Tests"
- git checkout class-based
- ps: Pop-Location
```

**Comment by randomnote1** [comment]
Looks like that change worked for pointing to the correct directory.

**Comment by randomnote1** [comment]
It just ran in R3 of #678 .

**Comment by johlju** [comment]
PR opened https://github.com/PowerShell/DscResource.Tests/pull/150.

**Comment by kwirkykat** [comment]
@johlju @randomnote1 
I got the class-based resources to work alongside mof-based resources in the DscResources folder back in June-ish? of 2016. I was trying it out because WindowsPackageCab is actually a class-based resource in PSDesiredStateConfiguration and we wanted to just insert it as it was into PSDscResources. I had WindowsPackageCab as the RootModule in the manifest, but the PS team said it would work in NestedModules too, but I didn't try that. Obviously something has changed since then to break class-based resources in the DscResources folder, but I'm not sure what. I can't get a straight answer from the team either on whether it was ever even supposed to work or not, so I wouldn't bet on anyone fixing it anytime soon. If the DscClassResources folder works, cool.

I struggled a lot with unit testing the class-based resources. It ended up sort-of working with the `using` keyword but I had a problem where I couldn't run the tests twice in a row. I think you guys are talking about a similar issue somewhere but without using that keyword. Our solution was the separate modules as described in https://github.com/PowerShell/DscResources/issues/208#issuecomment-255462702

But I had so much trouble with the testing that I ended up just converting WindowsPackageCab to a mof-based resource. And it sounds like that's what's happening with this resource too...

**Comment by randomnote1** [comment]
The testing worked great as long as it only ran once per PowerShell session. For whatever reason, it does not find the mocked commands the second time.

**Comment by johlju** [comment]
@kwirkykat Thanks for clarifying. Then we was on the right track!

Actually I think the resource works in either DscResource and DscClassResource, but depending om PowerShell-version, because @randomnote1 and I see different result. See review comments in PR #678. And the resource must be added to NestedModules, which works, to be able to support more than one resource (?). 

The biggest problem we have now, and the reason for converting to .mof, is that the tests is not trustworthy since they break when I try to add the markdown (or example) common tests in PR #681. And I think we see [this error](https://stackoverflow.com/questions/42838107/remove-class-from-memory-in-powershell/42878789#42878789).
So I took @randomnote1 tests for the class-based resource and broke it down to bare minimum step by step to see if it started working somewhere. But ended up with a mockup here; https://github.com/johlju/PesterTestClassBasedResource. Which in turn was posted to Pester repository as an issue here https://github.com/pester/Pester/issues/797. 

It turns out that the above mockup works to run several times in a row with PowerShell 6.0 beta 3. I think something happened in the alpha 14 release, because there is a mention of classes there.

But as you can see in the Pester issue, Joey has moved classes work to milestone 6.1, so I guess we can't look to using class-based resource until, at least, PowerShell 6.0 has been released. Because not until then they can be tested correctly. And hopefully things get even more stable when PowerShell 6.1 is release.

Therefor I thinking I will branch the current dev-branch to a new branch in this repository, referencing this issue, so we have something to test with once PowerShell 6.0 has been released and we can use that on a build worker.
